push/src/main/java/org/aerogear/mobile/push/PushService.java

```diff
@@ -0,0 +1,315 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import com.google.firebase.iid.FirebaseInstanceId;
+import com.google.firebase.messaging.FirebaseMessaging;
+
+import android.content.Context;
+import android.content.pm.ApplicationInfo;
+import android.content.pm.PackageManager;
+import android.os.Bundle;
+import android.util.Base64;
+
+import org.aerogear.mobile.core.Callback;
+import org.aerogear.mobile.core.MobileCore;
+import org.aerogear.mobile.core.ServiceModule;
+import org.aerogear.mobile.core.configuration.ServiceConfiguration;
+import org.aerogear.mobile.core.exception.HttpException;
+import org.aerogear.mobile.core.executor.AppExecutors;
+import org.aerogear.mobile.core.http.HttpRequest;
+import org.aerogear.mobile.core.http.HttpResponse;
+
+public class PushService implements ServiceModule {
+
+    public static final String DEFAULT_MESSAGE_HANDLER_KEY = "DEFAULT_MESSAGE_HANDLER_KEY";
+
+    private static final String registryDeviceEndpoint = "rest/registry/device";
+    private static final String JSON_ANDROID_CONFIG_KEY = "android";
+    private static final String JSON_VARIANT_ID_KEY = "variantId";
+    private static final String JSON_VARIANT_SECRET_KEY = "variantSecret";
+    private static final String JSON_SENDER_ID_KEY = "senderId";
+
+
+    private static final List<MessageHandler> MAIN_THREAD_HANDLERS =
+                    Collections.synchronizedList(new ArrayList<>());
+    private static final List<MessageHandler> BACKGROUND_THREAD_HANDLERS =
+                    Collections.synchronizedList(new ArrayList<>());
+
+    private final String deviceType = "ANDROID";
+    private final String operatingSystem = "android";
+    private final String osVersion = android.os.Build.VERSION.RELEASE;
+
+    private MobileCore core;
+    private String url;
+    private UnifiedPushCredentials unifiedPushCredentials;
+
+    private static MessageHandler defaultHandler;
+
+    @Override
+    public String type() {
+        return "push";
+    }
+
+    @Override
+    public void configure(MobileCore core, ServiceConfiguration serviceConfiguration) {
+        this.core = core;
+        this.url = serviceConfiguration.getUrl();
+        getDefaultHandler(core.getContext());
+
+        try {
+            JSONObject android = new JSONObject(
+                            serviceConfiguration.getProperties().get(JSON_ANDROID_CONFIG_KEY));
+
+            unifiedPushCredentials = new UnifiedPushCredentials();
+            unifiedPushCredentials.setVariant(android.getString(JSON_VARIANT_ID_KEY));
+            unifiedPushCredentials.setSecret(android.getString(JSON_VARIANT_SECRET_KEY));
+            unifiedPushCredentials.setSender(android.getString(JSON_SENDER_ID_KEY));
+
+        } catch (JSONException e) {
+            MobileCore.getLogger().error(e.getMessage(), e);
+            // Should never happen
```

Comment:
Should this rethrow as a runtime exception?
I think there may be a exception type in the project we can use as well.

Reply:
```java
throw new ConfigurationNotFoundException("An error occurred while trying to load the push config");
```

wdyt?